File: services/telemetry-service/src/main/java/net/blueshell/telemetry/mapping/TelemetryMapper.java
package net.blueshell.telemetry.mapping;

import net.blueshell.common.dto.TelemetryDTO;
import net.blueshell.common.mapper.BaseMapper;
import net.blueshell.telemetry.model.Telemetry;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring")
public abstract class TelemetryMapper extends BaseMapper<Telemetry, TelemetryDTO> {

    @Mapping(target = "deletedAt", ignore = true)
    public abstract Telemetry fromDTO(TelemetryDTO dto);

    public abstract TelemetryDTO toDTO(Telemetry telemetry);
}

File: api/src/main/kotlin/net/blueshell/api/telemetry/web/mapper/TelemetryMapper.kt
package net.blueshell.api.telemetry.web.mapper

import net.blueshell.api.telemetry.web.dto.TelemetryDTO
import net.blueshell.api.shared.mapper.BaseMapper
import net.blueshell.api.telemetry.persistence.Telemetry
import org.mapstruct.BeanMapping
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.MappingTarget

@Mapper(componentModel = "spring")
abstract class TelemetryMapper : BaseMapper<Telemetry, TelemetryDTO>() {
    @BeanMapping(ignoreByDefault = true)
    @Mapping(target = "url")
    @Mapping(target = "platform")
    @Mapping(target = "version")
    abstract fun fromDTO(dto: TelemetryDTO, @MappingTarget telemetry: Telemetry): Telemetry

    @BeanMapping(ignoreByDefault = true)
    @Mapping(target = "id")
    @Mapping(target = "url")
    @Mapping(target = "platform")
    @Mapping(target = "version")
    abstract override fun toDTO(telemetry: Telemetry): TelemetryDTO
}
